File: android/src/main/java/com/gbwisx/msal_mobile/Payloads.java
package com.gbwisx.msal_mobile;

import com.google.gson.annotations.SerializedName;

import com.microsoft.identity.client.IAccount;
import com.microsoft.identity.client.IAuthenticationResult;

import java.text.SimpleDateFormat;
import java.util.Map;

public class Payloads {
    static interface MsalMobileResultPayload {}

    static class GetAccountResultPayload implements MsalMobileResultPayload {
        @SerializedName("currentAccount")
        private Payloads.Account currentAccount;

        @SerializedName("accountLoaded")
        private boolean accountLoaded;

        GetAccountResultPayload(IAccount currentMsalAccount) {
            accountLoaded = true;
            if (currentMsalAccount != null) {
                currentAccount = new Payloads.Account(currentMsalAccount.getTenantId(), currentMsalAccount.getClaims(), currentMsalAccount.getAuthority(), currentMsalAccount.getId(), currentMsalAccount.getUsername());
            }
        }
    }

    static class Account {
        @SerializedName("tenantId")
        private String tenantId;

        @SerializedName("claims")
        private Map<String, ?> claims;

        @SerializedName("authority")
        private String authority;

        @SerializedName("id")
        private String id;

        @SerializedName("username")
        private String username;

        Account(String accountTenantId, Map<String, ?> accountClaims, String accountAuthority, String accountId, String accountUsername) {
            tenantId = accountTenantId;
            claims = accountClaims;
            authority = accountAuthority;
            id = accountId;
            username = accountUsername;
        }
    }

    static class AuthenticationResultPayload implements MsalMobileResultPayload {
        @SerializedName("cancelled")
        private boolean cancelled;

        @SerializedName("success")
        private boolean success;

        @SerializedName("accessToken")
        private String accessToken;

        @SerializedName("tenantId")
        private String tenantId;

        @SerializedName("scope")
        private String[] scope;

        @SerializedName("expiresOn")
        private String expiresOn;

        private AuthenticationResultPayload(final boolean authSuccessful, final boolean authCancelled, final String authAccessToken) {
            success = authSuccessful;
            cancelled = authCancelled;
            accessToken = authAccessToken;
        }

        private AuthenticationResultPayload(IAuthenticationResult result) {
            success = true;
            cancelled = false;
            accessToken = result.getAccessToken();
            tenantId = result.getTenantId();
            scope = result.getScope();

            SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
            expiresOn = formatter.format(result.getExpiresOn());
        }

        static AuthenticationResultPayload success(IAuthenticationResult result) {
            return new AuthenticationResultPayload(result);
        }

        static AuthenticationResultPayload cancelled() {
            return new AuthenticationResultPayload(false, true, null);
        }
    }
}

File: ios/Classes/Payloads.swift
//
//  Payloads.swift
//  msal_mobile
//
//  Created by Greg Burke on 12/28/19.
//

import Foundation
import MSAL

class MsalMobileResultPayload : Codable {
    private var currentAccount : Account?
    private var accountLoaded : Bool
    
    private var cancelled : Bool
    private var success : Bool
    private var accessToken : String?
    private var tenantId : String?
    private var scope : [String]
    private var expiresOn : String?
    
    init() {
        cancelled = false
        success = true
        accessToken = nil
        tenantId = nil
        scope = []
        expiresOn = nil
        
        currentAccount = nil
        accountLoaded = false
    }
    
    init(authSuccessful : Bool, authCancelled : Bool, authAccessToken : String?) {
        self.success = authSuccessful
        self.cancelled = authCancelled
        self.accessToken = authAccessToken
        self.tenantId = nil
        self.scope = []
        self.expiresOn = nil
        
        self.currentAccount = nil
        self.accountLoaded = false
    }
    
    init(msalResult : MSALResult) {
        self.cancelled = false
        self.success = true
        self.accessToken = msalResult.accessToken
        self.tenantId = msalResult.tenantProfile.tenantId
        self.scope = msalResult.scopes
        
        let dateFormatterPrint = DateFormatter()
        dateFormatterPrint.dateFormat = "yyyy-MM-dd HH:mm:ss"
        self.expiresOn = dateFormatterPrint.string(from: msalResult.expiresOn)
        
        self.currentAccount = nil
        self.accountLoaded = false
    }
    
    init(currentMsalAccount : MSALAccount) {
        self.accountLoaded = true
        self.currentAccount = Account(msalAccount: currentMsalAccount)
        
        self.cancelled = false
        self.success = true
        self.accessToken = nil
        self.tenantId = nil
        self.scope = []
        self.expiresOn = nil
    }
}

class Account : Codable {
    private var tenantId : String?
    private var claims : [String : String]?
    private var authority : String?
    private var id : String?
    private var username : String?
    
    init(msalAccount : MSALAccount) {
        let info = msalAccount.homeAccountId

        tenantId = info?.tenantId
        claims = flattenClaims(msalClaims: msalAccount.accountClaims)
        authority = nil // not available
        id = msalAccount.identifier
        username = msalAccount.identifier
    }
    
    private func flattenClaims(msalClaims: [String: Any]?) -> [String: String]? {
        if let claims = self.claims {
            return Dictionary(uniqueKeysWithValues: claims.map { key, value in (key, value as String) })
        }
        return nil
    }
}
